485 public datasets visible when not logged in

tcnichol opened the pull request:
ready for review now

lmarini reviewed (changes requested):
Haven't looked at the code yet, but wanted to write down a few comments on the functionality:

- Error downloading dataset from public page (works in private)
- File page doesn't show details component on right hand side
- Extraction history tab looks different from when logged in. Probably permission. Do we want to even show this? I am not sure a public users needs this info. Remove tab?

tcnichol commented:
> Haven't looked at the code yet, but wanted to write down a few comments on the functionality:
> 
> * Error downloading dataset from public page (works in private)
> * File page doesn't show details component on right hand side
> * Extraction history tab looks different from when logged in. Probably permission. Do we want to even show this? I am not sure a public users needs this info. Remove tab?

I have removed extraction history from the public views. The public file page should also show the details properly now.

With download, still looking for a fix. Noticed something strange - repeatedly clicking the download button will eventually download a public dataset while not logged in, but one click has no effect. Working on this.

tcnichol commented:
@lmarini i didn't do anything specifically to fix download, but it now appears to work on one click. The dataset I made public locally took a while, but I also realize I put some rather large files in it. If you can check again (and the other issues) let me know.

longshuicy commented on backend/app/routers/authorization.py:
```diff
@@ -1,3 +1,5 @@
+import datetime
```
any reason you need date time?

longshuicy commented on backend/app/models/files.py:
```diff
@@ -95,6 +109,7 @@ class FileDBViewList(View, FileBase):
     content_type: ContentType = ContentType()
     thumbnail_id: Optional[PydanticObjectId] = None
+    status: str = FileStatus.PRIVATE.name
 
```
you already have it in the "FileBase", maybe no need to repeat here unless it's necessary?

tcnichol replied:
I removed this, it was not needed.

longshuicy commented on backend/app/routers/datasets.py:
```diff
@@ -522,6 +530,12 @@ async def save_files(
+            authenticated = False
+            if dataset.status == "PUBLIC":
+                public = True
+            if dataset.status == "AUTHENTICATED":
```
Could you help me understand the concept of "AUTHENTICATED"?

tcnichol replied:
If a dataset is "public" then anyone can see it, even if they are not logged in.

If a dataset is "authenticated" then anyone with an account who is logged in can see it.

longshuicy replied:
thanks for the explanation!

longshuicy commented on backend/app/main.py:
```diff
@@ -164,6 +185,11 @@
     dependencies=[Depends(get_current_username)],
 )
+api_router.include_router(
+    elasticsearch.router,
```
I see you have `public_elasticsearch.py` file, maybe it's a typo?

longshuicy commented on backend/app/main.py:
```diff
@@ -182,6 +208,11 @@
     dependencies=[Depends(get_current_username)],
 )
+api_router.include_router(
+    visualization.router,
```
same did you mean `public_visualization`?

longshuicy commented on backend/app/routers/public_datasets.py:
```diff
@@ -0,0 +1,353 @@
+import datetime
```
Just curious is there any similarity to `datasets.py`?

tcnichol replied:
it has some of the methods of datasets but not the Depends(get_current_user) and it does not have a dependency added to the router in main.

It only has the get methods.

longshuicy commented on frontend/src/routes.tsx:
```diff
@@ -98,6 +101,12 @@ export const AppRoutes = (): JSX.Element => {
 	return (
 		<BrowserRouter>
 			<Routes>
+				<Route
```
did we settle on using a separate route of /public?

tcnichol replied:
Yes, this one does not have the Depends(get_current_user) for methods but will always only return datasets that are public, and does not have routes that edit or create datasets. 

I did think that it might have been a good idea to have made a dependency that checks public, but it didn't seem absolutely necesary.

tcnichol replied:
I just pushed a change that I thought would work for this, but so far it still redirects to keycloak. Working on this.

tcnichol replied:
I think this is fixed now.

longshuicy commented on frontend/src/utils/common.js:
```diff
@@ -52,6 +52,25 @@ export async function downloadResource(url) {
 	}
 }
 
+export async function downloadPublicResource(url) {
```
could we just reuse the downloadResource function? the content is exactly the same except leave out the "authHeader"

tcnichol replied:
I tested and I could use downloadResource, removed the other

tcnichol commented on .run/uvicorn.run.xml:
```diff
@@ -7,8 +7,9 @@
+    <option name="SDK_NAME" value="Pipenv (clowder2)" />
     <option name="WORKING_DIRECTORY" value="$PROJECT_DIR$/backend" />
-    <option name="IS_MODULE_SDK" value="true" />
+    <option name="IS_MODULE_SDK" value="false" />
```
reverted.

tcnichol commented on backend/app/deps/authorization_deps.py:
```diff
@@ -227,7 +254,15 @@ async def __call__(
+                    file_id, "file", self.role, current_user
+                )
+                if public_access:
+                    return True
```
I did it like this since it looked like the method was supposed to either return True or throw an error. But I did change it to returning public_access instead of 'True' since that looked confusing.

tcnichol commented on backend/app/deps/authorization_deps.py:
```diff
@@ -6,13 +6,40 @@
+    resource_id: str,
+    resource_type: str,
+    role: RoleType,
+    current_user=Depends(get_current_username),
```
Let

tcnichol replied:
Let me do some testing here.

tcnichol replied:
The reason we have the user is that if the user is logged in, they should still see PUBLIC or AUTHENTICATED datasets. That's why it was done like this.

tcnichol commented on backend/app/deps/authorization_deps.py:
```diff
@@ -6,13 +6,40 @@
+    current_user=Depends(get_current_username),
+) -> bool:
+    has_public_access = False
+    if role == RoleType.VIEWER:
```
The method is only called in cases where the user did not have a specific role on a dataset.

If the role being requested was not Viewer, check_public_access should return False, even if the dataset status is 'public.' 

Passing in the role might be confusing, so I might try to think of a better way to handle this. I was just going for a method that would be easy to reuse.

tcnichol replied:
I'm going to leave this for now, I cannot think of a better way.

tcnichol commented:
Some newer changes.

In frontend/components/visualizations there is logic that checks if we are looking at a public component, and then uses a different route to download the file and get visualizations depending. The visualizations were not showing up if a user was not logged in.

tcnichol commented on backend/app/deps/authorization_deps.py:
```diff
@@ -41,24 +73,11 @@ async def get_role_by_file(
+        )
+        if authorization is None and public_access:
+            return RoleType.VIEWER
         return authorization.role
```
Yes, that is changed. It looks like it's supposed to either return a role or throw an error, not return None

longshuicy replied:
There is still some quirks here. For file You might still want to catch the case that "authorization" is None but public_access is "false"

I got an error that traced to this place. Try: 
- mark an empty dataset "public"
- upload a file to that dataset.
- go to that file
- I get an error 
```
 File "/Users/cwang138/.local/share/virtualenvs/backend-pkZ6g-7g/lib/python3.9/site-packages/fastapi/dependencies/utils.py", line 622, in solve_dependencies
    solved = await call(**sub_values)
  File "/Users/cwang138/Documents/Projects/Clowder/clowder2/backend/app/deps/authorization_deps.py", line 81, in get_role_by_file
    return authorization.role
AttributeError: 'NoneType' object has no attribute 'role'
```

longshuicy replied:
<img width="523" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/5d16ac24-fa88-44fd-9130-17abf417939d">

tcnichol replied:
@longshuicy I could not replicate this, but I did add throwing an exception. Let me know if this works.

longshuicy commented:
<img width="1548" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/a0d1a4f6-dff5-41a2-ad8b-67e14dcf95a7">

The "Search" on the left drawer doesn't work. I think search is a protected endpoint. Do we want to remove that altogether?

longshuicy commented:
Same as the search bar on top
<img width="706" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/0dc71a27-20de-4640-84ff-f1dec4c0533c">

longshuicy reviewed (changes requested):
<img width="885" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/5218f989-6cff-451d-a0a0-cb50ac3cee94">
 Metadata is protected as well

longshuicy commented on frontend/src/actions/public_dataset.js:
```diff
@@ -0,0 +1,144 @@
+export const RECEIVE_PUBLIC_DATASETS = "RECEIVE_PUBLIC_DATASETS";
+
+export function fetchPublicDatasets(skip = 0, limit = 21) {
+	console.log("we are here");
```
Could you remove the console.log?

tcnichol replied:
removed

longshuicy commented on frontend/src/components/visualizations/Image/Image.tsx:
```diff
@@ -2,25 +2,37 @@ import React, { useEffect, useState } from "react";
+				downloadUrl = generateFileDownloadUrl(fileId, 0);
+			}
 		}
+		console.log('download URL is', downloadUrl);
```
Could you remove the console.log?

tcnichol replied:
removed

longshuicy commented on frontend/src/components/visualizations/PublicVisualization.tsx:
```diff
@@ -0,0 +1,184 @@
+			getPublicVisConfig(datasetId);
+		}
+	}, [fileId, datasetId]);
+	console.log("file summary is", fileSummary);
```
Could you remove the console.log?

tcnichol replied:
removed

longshuicy commented on frontend/src/components/visualizations/PublicVisualizationCard.tsx:
```diff
@@ -0,0 +1,90 @@
+	const { publicVisComponentDefinition, publicVisualizationDataItem } = props;
+	const [expanded, setExpanded] = React.useState(false);
+	const [visShareModalOpen, setVisShareModalOpen] = useState(false);
+	console.log("public visualization card");
```
Could you remove the console.log?

tcnichol replied:
removed

longshuicy commented on frontend/src/utils/common.js:
```diff
@@ -31,6 +31,26 @@ export function getHeader() {
 
+export async function downloadPublicResource(url) {
+	const authHeader = getHeader();
+	console.log('downloading public resource', url);
```
Could you remove the console.log?

tcnichol replied:
removed

longshuicy commented on frontend/src/components/files/FilesTableFileEntry.tsx:
```diff
@@ -16,13 +16,14 @@ type FilesTableFileEntryProps = {
 	const [thumbnailUrl, setThumbnailUrl] = useState("");
 	const [selectedVersion, setSelectedVersion] = useState(file.version_num);
-
+	console.log('the public view is', publicView);
```
Could you remove the console.log?

tcnichol replied:
removed

longshuicy commented:
After reading your code, I realize you probably need to point those frontend link/buttons to the "public_xxx" endpoint, e.g. search, metadata, etc. Could you double check if there is anything else missing?

ddey2 commented:
Few things I noticed:

1. Search is working now. 
2. I can upload file to a public dataset. 
3. If I try to view 'User metadata' tab, I get error. 
<img width="1722" alt="Screenshot 2024-01-16 at 2 36 28 PM" src="https://github.com/clowder-framework/clowder2/assets/22405928/3658be30-5025-4884-b31f-00ad5dc7541f">
4. If I got to 'sharing tab', I get 'Forbidden page' error. Sharing doesn't work.

ddey2 reviewed (changes requested):
Left the things I noted in comment

lmarini commented:
@tcnichol can you update the description of the PR to describe what it does? Not just for the current reviewers but for future reference as well. Thank you.

tcnichol commented:
How to test this pull request.

Note - the new 'default' page is now the public page. Login and register are in the upper right hand corner.

Create at least 2 Users. 

Using User1, create datasets and upload files. Then mark at least one dataset 'public' by changing the status from 'private' to 'public.' 

If you log out as that user, then you will see the public datasets. Other users will also see them when logged in. So if User1 made a dataset public, User2 will see it when logged in.

tcnichol commented:
New instructions for testing.

If you start up a clowder instance, run the populate fake data script. It will generate some public datasets and they should be visible when you are not logged in.

The new default page is the public page, with login and register in the upper right corner.

longshuicy commented on frontend/src/actions/public_file.js:
```diff
@@ -0,0 +1,209 @@
+export const RECEIVE_PUBLIC_FILE_METADATA_JSONLD = "RECEIVE_PUBLIC_FILE_METADATA_JSONLD";
+
+export function fetchPublicFileMetadataJsonld(id) {
+	const url = `${config.hostname}/public/files/${id}/metadata.jsonld`;
```
Hmm do we even have a jsonld endpoint? you probably can leave this out

tcnichol replied:
removed.

longshuicy commented on frontend/src/actions/public_file.js:
```diff
@@ -0,0 +1,209 @@
+
+export function fetchPublicFilePreviews(id) {
+	const url = `${config.hostname}/public/files/${id}/getPreviews`;
+	return (dispatch) => {
```
same here we don't have endpoint for that. That's most likely just for v1 clowder

tcnichol replied:
removed.

longshuicy reviewed (changes requested):
### Could we remove those dividers?
<img width="261" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/0b4beeff-d2cd-4e43-9b24-4a94a8ae174d">
---
### Download version history doesn't work. i think the url is wrong, instead of public/files you should use public_files in the URL
<img width="827" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/4d27ec04-c975-4982-a6b5-69f572a165c0">
<img width="804" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/76b5267e-5876-432e-b93c-449b915df743">
---
### Download dataset button doesn't work. It's still pointing to the private endpoint.
<img width="443" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/79a77bcd-d6b1-4369-b96a-4409cede5cfc">
<img width="889" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/f1c97a3e-c43a-4291-9357-68b897887eae">
---
### Download dataset button inside the dataset page doesn't work either. Not sure why
<img width="1785" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/37d92a21-08b1-4668-89bb-f6af79dd7f4b">
<img width="967" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/015b2fd9-4386-4321-b5bb-dced417bc7a7">

tcnichol commented on frontend/src/actions/public_file.js:
```diff
@@ -0,0 +1,209 @@
+
+export function fetchPublicFileExtractedMetadata(id) {
+	const url = `${config.hostname}/api/v2/public/files/${id}/metadata`;
+	return (dispatch) => {
```
I tried to replicate what was done in actions/file.js with this one. Maybe both should be changed?

longshuicy replied:
I dont think this is being used anywhere. Maybe just legacy code
<img width="776" alt="image" src="https://github.com/clowder-framework/clowder2/assets/13950475/87f059e2-7a8d-4140-a4b9-1f1e3a488413">

tcnichol commented:
@longshuicy i removed the divider and addressed your comments.
There is one method that i left a comment on - in public_file I modeled the metadata function that gets extracted metadata after the one in actions/file.js. Maybe both should change?

longshuicy reviewed (approved):
All the suggestions are resolved now. Approve. 
But could you fix the black linting before merging?

tcnichol commented:
I'm not sure what is going wrong.

I am running

`pipenv run black develop`

and i see that it's changing populate_fake_data.py

but when I go to commit it shows nothing. 

Same with the backend/app
